### PR TITLE
Server capabilities managed better, user can disable formatting from preferences

### LINKED ISSUE #104: Add preference to enable/disable formatting support
Since we keep finding bugs in the formatting support, it'd be nice to provide support for a format.enabled=false preference.

Formatting support should be (un)registered dynamically. For reference, this is how jdt.ls handles dynamic registration : https://github.com/eclipse/eclipse.jdt.ls/blob/5b3ba86b681cd03e707c44347e4605a4fe1d4cda/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/InitHandler.java#L138-L140
and 
https://github.com/eclipse/eclipse.jdt.ls/blob/5b3ba86b681cd03e707c44347e4605a4fe1d4cda/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java#L226-L232

Of course, other features should be dynamically triggered, but let's start with formatting

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Fixes #104

### CODE REVIEW

**Comment by angelozerr** [comment]
@fbricon I trust you, please merge this PR if you think it's good. I hope it will continue to work with LSP4e.

File: .vscode/launch.json [hunk 1]
```diff
@@ -4,12 +4,13 @@
...
+
     {
       "type": "java",
       "name": "Debug (Attach)",
       "request": "attach",
       "hostName": "localhost",
-      "port": 1054
-    }
+	  "port": 1054
+	}
```
Comment (fbricon) [inline .vscode/launch.json]: unnecessary changes

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java [hunk 2]
```diff
@@ -106,11 +157,70 @@ public void updateSettings(Object initializationOptionsSettings) {
...
+		if (capabilityManager.getClientCapabilitiesWrapper().isRangeFormattingDynamicRegistrationSupported()) {
+			toggleCapability(clientPreferences.getBoolean(FORMATTING_PREFERENCE), Preferences.FORMATTING_RANGE_ID, 
+				Preferences.TEXT_DOCUMENT_RANGE_FORMATTING, null);
+		}
+	}
+
+	/**
+	 * 
+	 */
+	private void toggleCapability(boolean enabled, String id, String capability, Object options) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java]: belongs to capabilityManager

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java [hunk 3]
```diff
@@ -106,11 +157,70 @@ public void updateSettings(Object initializationOptionsSettings) {
...
+		if (enabled) {
+			registerCapability(id, capability, options);
+		} else {
+			unregisterCapability(id, capability);
+		}
+	}
+
+	public void unregisterCapability(String id, String method) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java]: belongs to capabilityManager

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java [hunk 4]
```diff
@@ -55,33 +64,74 @@
 	private LanguageClient languageClient;
 	private final ScheduledExecutorService delayer;
 	private Integer parentProcessId;
+	private XMLCapabilityManager capabilityManager;
+
+	private Set<String> registeredCapabilities = new HashSet<>(3);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java]: belongs to capabilityManager

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java [hunk 5]
```diff
@@ -55,33 +64,74 @@
...
+	}
+
+	/*
+     * Registers all capabilities that do not support client side preferences to turn on/off
+	 *
+	 * (non-Javadoc)
+	 * @see org.eclipse.lsp4j.services.LanguageServer#initialized(org.eclipse.lsp4j.InitializedParams)
+	 */
+	@Override
+	public void initialized(InitializedParams params) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java]: delegate to capabilityManager.initializeCapabilities()

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java [hunk 6]
```diff
@@ -106,11 +157,70 @@ public void updateSettings(Object initializationOptionsSettings) {
...
+	private void updateClientCapabilityPreferences(ClientCapabilityPreferences updatedCapabilities) {
+		this.capabilityManager.getClientCapabilityPreferences().merge(updatedCapabilities);	
+	}
+
+	/**
+	 * Registers all capabilities that this server can support client side preferences to turn on/off
+	 * 
+	 * If a capability is not dynamic, it's handled by {@link InitialServerCapabilities}
+	 */
+	public void syncDynamicCapabilitiesWithPreferences() {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/XMLLanguageServer.java]: move to capabilityManager

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/AllCapabilities.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package org.eclipse.lsp4xml.settings.capabilities;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/AllCapabilities.java]: missing copyright header

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
+package org.eclipse.lsp4xml.settings.capabilities;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java]: missing copyright header

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * Determines if a client supports a specific capability dynamically
+ */
+public class ClientCapabilitiesWrapper {
+	private boolean dynamicRegistrationSupported;
+
+	public ClientCapabilities capabilities;
+
+	public ClientCapabilitiesWrapper(ClientCapabilities capabilities, boolean dynamicRegistrationSupported) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java]: not sure if capabilities can be null, depending on the client, so you might need to add some null-checks

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		this.putBoolean(LINK_PREFERENCE, true);
+		this.putBoolean(CODE_ACTION_PREFERENCE, true);
+	}
+
+
+	public ClientCapabilityPreferences merge(ClientCapabilityPreferences capabilities) {
+		
+		capabilities.entrySet().stream().forEach(entry -> {//
+			this.put(entry.getKey(), entry.getValue()); //
+		LOGGER.info("'" + entry.getKey() + "' set to " + entry.getValue());
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java]: indentation

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
+package org.eclipse.lsp4xml.settings.capabilities;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java]: missing copyright header

**Comment by angelozerr** [comment]
Guys could you just explain me what is teh benefit to use dynamic capability (speed the start of server?). Thanks!

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/Preferences.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
+package org.eclipse.lsp4xml.settings.capabilities;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/Preferences.java]: missing copyright header

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
+package org.eclipse.lsp4xml.settings.capabilities;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: missing copyright header

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+	public void setClientCapabilityPreferences(ClientCapabilityPreferences clientCapabilityPreferences) {
+		this.clientCapabilityPreferences = clientCapabilityPreferences;
+	}
+
+	public ClientCapabilityPreferences getClientCapabilityPreferences() {
+		return this.clientCapabilityPreferences;
+	}
+
+	public void updateClientCapabilitiesWrapper(ClientCapabilities clientCapabilities) {
+		boolean dynamicRegistrationSupported = clientCapabilities.getTextDocument() != null;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: calling it dynamicRegistrationSupported is a bit dubious. Not sure about that one.

**Comment by fbricon** [comment]
@angelozerr this allows the server to register capabilities conditionally. eg. if a user sets some preference to disable formatting, then the server will unregister the formatting capability and the client will no longer display the format menu.
If code lenses are disabled by preferences, the server will unregister codelens support and the client will stop sending codelens requests. And so on and so forth.

**Comment by NikolasKomonen** [comment]
@fbricon updated

**Comment by NikolasKomonen** [comment]
@fbricon updated

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
+/**
+ *  Copyright (c) 2018 Angelo ZERR
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java]: Copyright (c) 2018 Red Hat, Inc. and others.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/AllCapabilities.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
+/**
+ *  Copyright (c) 2018 Angelo ZERR
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/AllCapabilities.java]: Copyright (c) 2018 Red Hat, Inc. and others.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+		this.v3Supported = v3Supported;
+	}
+
+	/**
+	 * 			IMPORTANT
+	 * 
+	 * This should be up to date with all Server supported capabilities
+	 * 
+	 */
+
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java]: can replace capabilities.getTextDocument() with getTextDocument()

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+	
+	private boolean isDynamicRegistrationSupported(DynamicRegistrationCapabilities capability) {
+		return capability != null && capability.getDynamicRegistration().booleanValue();
+	}
+
+	public TextDocumentClientCapabilities getTextDocument() {
+		return this.capabilities.getTextDocument();
+	}
+
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java]: remove empty lines

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java [hunk 19]
```diff
@@ -0,0 +1,87 @@
+/**
+ *  Copyright (c) 2018 Angelo ZERR
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java]: Copyright (c) 2018 Red Hat, Inc. and others.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java [hunk 20]
```diff
@@ -0,0 +1,74 @@
+/**
+ *  Copyright (c) 2018 Angelo ZERR
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java]: Copyright (c) 2018 Red Hat, Inc. and others.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/Preferences.java [hunk 21]
```diff
@@ -0,0 +1,65 @@
+/**
+ *  Copyright (c) 2018 Angelo ZERR
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/Preferences.java]: Copyright (c) 2018 Red Hat, Inc. and others.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 22]
```diff
@@ -0,0 +1,144 @@
+/**
+ *  Copyright (c) 2018 Angelo ZERR
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: Copyright (c) 2018 Red Hat, Inc. and others.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+	public void setClientCapabilityPreferences(ClientCapabilityPreferences clientCapabilityPreferences) {
+		this.clientCapabilityPreferences = clientCapabilityPreferences;
+	}
+
+	public ClientCapabilityPreferences getClientCapabilityPreferences() {
+		return this.clientCapabilityPreferences;
+	}
+
+	public void updateClientCapabilitiesWrapper(ClientCapabilities clientCapabilities) {
+		boolean v3Supported = clientCapabilities.getTextDocument() != null;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: can be moved to ClientCapabilitiesWrapper constructor

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * Determines if a client supports a specific capability dynamically
+ */
+public class ClientCapabilitiesWrapper {
+	private boolean v3Supported;
+
+	public ClientCapabilities capabilities;
+
+	public ClientCapabilitiesWrapper(ClientCapabilities capabilities, boolean v3Supported) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilitiesWrapper.java]: v3Supported can be inferred directly inside this constructor, from capabilities

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/AllCapabilities.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+ *  Contributors:
+ *  Red Hat Inc. - initial API and implementation
+ */
+
+package org.eclipse.lsp4xml.settings.capabilities;
+
+/**
+ * AllCapabilities
+ */
+public class AllCapabilities {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/AllCapabilities.java]: this class seems unused

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  which accompanies this distribution, and is available at
+ *  http://www.eclipse.org/legal/epl-v10.html
+ *
+ *  Contributors:
+ *  Red Hat Inc. - initial API and implementation
+ */
+package org.eclipse.lsp4xml.settings.capabilities;
+
+import static org.eclipse.lsp4xml.settings.capabilities.ClientCapabilityPreferences.FORMATTING_PREFERENCE;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: unused import

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+
+	public void setTextDocumentService(XMLTextDocumentService xmlTextDocumentService) {
+		this.textDocumentService = xmlTextDocumentService;
+	}
+
+	public XMLTextDocumentService getTextDocumentService() {
+		return this.textDocumentService;
+	}
+
+	// public void setClientCapabilityPreferences(ClientCapabilityPreferences clientCapabilityPreferences) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: please remove commented code

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+import org.eclipse.lsp4j.ClientCapabilities;
+import org.eclipse.lsp4j.Registration;
+import org.eclipse.lsp4j.RegistrationParams;
+import org.eclipse.lsp4j.Unregistration;
+import org.eclipse.lsp4j.UnregistrationParams;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4xml.XMLTextDocumentService;
+import org.eclipse.lsp4xml.settings.XMLFormattingOptions;
+/**
+ * ClientPreferences
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: Please update javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+ */
+public class XMLCapabilityManager {
+
+	//ClientCapabilityPreferences clientCapabilityPreferences;
+	ClientCapabilitiesWrapper clientWrapper;
+	private Set<String> registeredCapabilities = new HashSet<>(3);
+	private LanguageClient languageClient;
+	private XMLTextDocumentService textDocumentService;
+
+	public XMLCapabilityManager(LanguageClient languageClient) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: XMLCapabilityManager(LanguageClient languageClient, XMLTextDocumentService textDocumentService)

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+		this.languageClient = languageClient;
+	}
+
+	public void setTextDocumentService(XMLTextDocumentService xmlTextDocumentService) {
+		this.textDocumentService = xmlTextDocumentService;
+	}
+
+	public XMLTextDocumentService getTextDocumentService() {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: unused, should be removed

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+	ClientCapabilitiesWrapper clientWrapper;
+	private Set<String> registeredCapabilities = new HashSet<>(3);
+	private LanguageClient languageClient;
+	private XMLTextDocumentService textDocumentService;
+
+	public XMLCapabilityManager(LanguageClient languageClient) {
+		this.languageClient = languageClient;
+	}
+
+	public void setTextDocumentService(XMLTextDocumentService xmlTextDocumentService) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: should be removed as service should be set in constructor

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4xml.XMLTextDocumentService;
+import org.eclipse.lsp4xml.settings.XMLFormattingOptions;
+/**
+ * ClientPreferences
+ */
+public class XMLCapabilityManager {
+
+	//ClientCapabilityPreferences clientCapabilityPreferences;
+	ClientCapabilitiesWrapper clientWrapper;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: private

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+import org.eclipse.lsp4j.UnregistrationParams;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4xml.XMLTextDocumentService;
+import org.eclipse.lsp4xml.settings.XMLFormattingOptions;
+/**
+ * ClientPreferences
+ */
+public class XMLCapabilityManager {
+
+	//ClientCapabilityPreferences clientCapabilityPreferences;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: remove commented code

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 34]
```diff
@@ -0,0 +1,157 @@
...
+	}
+
+	public ClientCapabilitiesWrapper getClientCapabilitiesWrapper() {
+		if(this.clientWrapper == null) {
+			this.clientWrapper = new ClientCapabilitiesWrapper();
+		}
+		return this.clientWrapper;
+	}
+
+	/**
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: remove empty javadoc, or write the doc, your choice ;-)

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java [hunk 35]
```diff
@@ -0,0 +1,74 @@
...
+	public static CompletionOptions DEFAULT_COMPLETION_OPTIONS = new CompletionOptions(false, Arrays.asList(".", ":", "<", "\"", "=", "/"));
+	public static TextDocumentSyncKind DEFAULT_SYNC_OPTION = TextDocumentSyncKind.Full;
+	public static DocumentLinkOptions DEFAULT_LINK_OPTIONS = new DocumentLinkOptions(true);
+
+	/**
+	 * Returns all default server capabilities, unless a dynamic option is selected
+	 * @param manager
+	 * @return
+	 */
+	public static ServerCapabilities getNonDynamicServerCapabilities(XMLCapabilityManager manager) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java]: getNonDynamicServerCapabilities(ClientCapabilitiesWrapper clientCapabilities)

We just don't need a XMLCapabilityManager here

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java [hunk 36]
```diff
@@ -0,0 +1,74 @@
...
+
+import org.eclipse.lsp4j.CompletionOptions;
+import org.eclipse.lsp4j.DocumentLinkOptions;
+import org.eclipse.lsp4j.ServerCapabilities;
+import org.eclipse.lsp4j.TextDocumentSyncKind;
+
+/**
+ * All default capabilities of this server
+ */
+public class InitialServerCapabilities {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java]: ServerCapabilitiesInitializer

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java [hunk 37]
```diff
@@ -0,0 +1,74 @@
...
+
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java]: use private constructor, as it only contains static methods

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java [hunk 38]
```diff
@@ -0,0 +1,74 @@
...
+	public static CompletionOptions DEFAULT_COMPLETION_OPTIONS = new CompletionOptions(false, Arrays.asList(".", ":", "<", "\"", "=", "/"));
+	public static TextDocumentSyncKind DEFAULT_SYNC_OPTION = TextDocumentSyncKind.Full;
+	public static DocumentLinkOptions DEFAULT_LINK_OPTIONS = new DocumentLinkOptions(true);
+
+	/**
+	 * Returns all default server capabilities, unless a dynamic option is selected
+	 * @param manager
+	 * @return
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java]: fix javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 39]
```diff
@@ -0,0 +1,157 @@
...
+		if (registeredCapabilities.add(id)) {
+			Registration registration = new Registration(id, method, options);
+			RegistrationParams registrationParams = new RegistrationParams(Collections.singletonList(registration));
+			languageClient.registerCapability(registrationParams);
+		}
+	}
+
+	public void initializeCapabilities() {
+		if (this.getClientCapabilitiesWrapper().isCodeActionDynamicRegistered()) {
+			registerCapability(Preferences.CODE_ACTION_ID, Preferences.TEXT_DOCUMENT_CODE_ACTION, null);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: import static org.eclipse.lsp4xml.settings.capabilities.ServerCapabilitiesConstants.*; so you don't have to add the class prefix before each constant

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java [hunk 40]
```diff
@@ -0,0 +1,74 @@
...
+import org.eclipse.lsp4j.TextDocumentSyncKind;
+
+/**
+ * All default capabilities of this server
+ */
+public class InitialServerCapabilities {
+
+	public static CompletionOptions DEFAULT_COMPLETION_OPTIONS = new CompletionOptions(false, Arrays.asList(".", ":", "<", "\"", "=", "/"));
+	public static TextDocumentSyncKind DEFAULT_SYNC_OPTION = TextDocumentSyncKind.Full;
+	public static DocumentLinkOptions DEFAULT_LINK_OPTIONS = new DocumentLinkOptions(true);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/InitialServerCapabilities.java]: move 3 constants to ServerCapabilitiesConstants

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java [hunk 41]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * Holds all of the boolean preferences of each capability
+ * 
+ * Defaults to true if not assigned
+ */
+public class ClientCapabilityPreferences extends LinkedHashMap<String, Boolean> {
+
+  private static Logger LOGGER = Logger.getLogger(ClientCapabilityPreferences.class.getName());
+  // private LinkedHashMap<String, Boolean> capabilityMap = new
+  // LinkedHashMap<String, Boolean>();
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java]: delete commented code

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/Line.java [hunk 42]
```diff
@@ -12,58 +12,56 @@
 package org.eclipse.lsp4xml.commons;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/commons/Line.java]: I don't see useful changes here. file should be dropped from commit

**Comment by NikolasKomonen** [comment]
@fbricon updated

File: .vscode/launch.json [hunk 43]
```diff
@@ -3,13 +3,11 @@
   // Hover to view descriptions of existing attributes.
```
Comment (fbricon) [inline .vscode/launch.json]: unnecessary change

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLClientSettings.java [hunk 44]
```diff
@@ -22,6 +23,8 @@
 
 	private XMLFormattingOptions format;
 
+	private ClientCapabilityPreferences capabilities;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLClientSettings.java]: this is effectively unused

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLClientSettings.java [hunk 45]
```diff
@@ -38,7 +41,19 @@ public XMLFormattingOptions getFormat() {
 		return format;
 	}
 
+	public void setCapabilities(ClientCapabilityPreferences capabilities) {
+		this.capabilities = capabilities;
+	}
+
+	public ClientCapabilityPreferences getCapabilities() {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/XMLClientSettings.java]: nobody calls this method

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java [hunk 46]
```diff
@@ -0,0 +1,85 @@
...
+package org.eclipse.lsp4xml.settings.capabilities;
+
+import java.util.LinkedHashMap;
+import java.util.logging.Logger;
+
+/**
+ * Holds all of the boolean preferences of each capability
+ * 
+ * Defaults to true if not assigned
+ */
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ClientCapabilityPreferences.java]: This class is not used anywhere, please remove it

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ServerCapabilitiesConstants.java [hunk 47]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.Arrays;
+import java.util.UUID;
+
+import org.eclipse.lsp4j.CompletionOptions;
+import org.eclipse.lsp4j.DocumentLinkOptions;
+import org.eclipse.lsp4j.TextDocumentSyncKind;
+
+/**
+ * Preferences
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ServerCapabilitiesConstants.java]: fix javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java [hunk 48]
```diff
@@ -0,0 +1,141 @@
...
+import org.eclipse.lsp4j.Unregistration;
+import org.eclipse.lsp4j.UnregistrationParams;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4xml.XMLTextDocumentService;
+import org.eclipse.lsp4xml.settings.XMLFormattingOptions;
+import static org.eclipse.lsp4xml.settings.capabilities.ServerCapabilitiesConstants.*;
+/**
+ * Manager for capability related tasks
+ */
+public class XMLCapabilityManager {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/XMLCapabilityManager.java]: Would be nice to have some unit tests for this class

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ServerCapabilitiesInitializer.java [hunk 49]
```diff
@@ -0,0 +1,72 @@
...
+
+	private ServerCapabilitiesInitializer() {
+	}
+
+	/**
+	 * Returns all default server capabilities that aren't dynamic
+	 * @param clientCapabilities
+	 * @return
+	 */
+	public static ServerCapabilities getNonDynamicServerCapabilities(ClientCapabilitiesWrapper clientCapabilities) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/settings/capabilities/ServerCapabilitiesInitializer.java]: unit tests please

**Comment by fbricon** [comment]
@NikolasKomonen please rebase against master

**Comment by NikolasKomonen** [comment]
@fbricon rebased
